kubectl force apply SIGSEGV with podDistributionBudgets

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

When trying to forcefully update `podDistributionBudget` using `kubectl apply --force -f pdb.json` command segmentatio voliation happens.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x15dcc0b]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/kubectl/resource.(*Helper).DeleteWithOptions(0xc420a7e200, 0xc4201cc95c, 0x4, 0xc4201cc92c, 0x4, 0x0, 0xc420a7e200, 0x20933ff)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/helper.go:103 +0x3b
k8s.io/kubernetes/pkg/kubectl/resource.(*Helper).Delete(0xc420a7e200, 0xc4201cc95c, 0x4, 0xc4201cc92c, 0x4, 0xc420a1e700, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/helper.go:99 +0x5c
k8s.io/kubernetes/pkg/kubectl/cmd.runDelete(0xc4201cc95c, 0x4, 0xc4201cc92c, 0x4, 0xc42026d260, 0x0, 0x0, 0xc420471900, 0x1, 0xffffffffffffffff, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:523 +0x221
k8s.io/kubernetes/pkg/kubectl/cmd.(*patcher).delete(0xc42026d960, 0xc4201cc95c, 0x4, 0xc4201cc92c, 0x4, 0x1a6, 0xc4206aef70)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:536 +0x9c
k8s.io/kubernetes/pkg/kubectl/cmd.(*patcher).deleteAndCreate(0xc42026d960, 0xc42094b500, 0x2e7, 0x2e7, 0xc4201cc95c, 0x4, 0xc4201cc92c, 0x4, 0xc4201cc95c, 0x4, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:623 +0x63
k8s.io/kubernetes/pkg/kubectl/cmd.(*patcher).patch(0xc42026d960, 0x2e41da0, 0xc420479a48, 0xc42094b500, 0x2e7, 0x2e7, 0x208d32a, 0x5, 0xc4201cc95c, 0x4, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:617 +0x43a
k8s.io/kubernetes/pkg/kubectl/cmd.RunApply.func1(0xc4201d4c00, 0x0, 0x0, 0x9, 0x15e6501)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:320 +0x76a
k8s.io/kubernetes/pkg/kubectl/resource.ContinueOnErrorVisitor.Visit.func1(0xc4201d4c00, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:348 +0x153
k8s.io/kubernetes/pkg/kubectl/resource.DecoratedVisitor.Visit.func1(0xc4201d4c00, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:324 +0xdf
k8s.io/kubernetes/pkg/kubectl/resource.FlattenListVisitor.Visit.func1(0xc4201d4c00, 0x0, 0x0, 0x0, 0x0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:389 +0x572
k8s.io/kubernetes/pkg/kubectl/resource.EagerVisitorList.Visit.func1(0xc4201d4c00, 0x0, 0x0, 0x1c0, 0x208d32a)

...
```

**What you expected to happen**:

podDistributionBudget to be replaced

**How to reproduce it (as minimally and precisely as possible)**:

create `pdb.json`:

```
{
  "apiVersion": "policy/v1beta1",
  "kind": "PodDisruptionBudget",
  "metadata": {
    "name": "test"
  },
  "spec": {
    "minAvailable": "60%",
    "selector": {
      "matchLabels": {
        "name": "myapp"
      }
    }
  }
}
```

Create `podDistributionBudget`:

```
kubectl apply -f pdb.json
```

Forcefull update (replace) `podDistributionBudget`:

```
kubectl apply -f pdb.json --force
```

**Anything else we need to know?**:

The same error happens on kubectl version 1.6 or 1.7. Manually deleting and recreating `podDistributionBudget` works.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7
- Cloud provider or hardware configuration**: gke
- OS (e.g. from /etc/os-release): Debian GNU/Linux 7 (wheezy)
- Kernel (e.g. `uname -a`): 3.16.0-4-amd64

This issue not only applies to PBP's but NetworkPolicy (and perhaps others).
```
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: access-nginx
spec:
  podSelector:
    matchLabels:
      run: nginx
  ingress:
  - from:
    - podSelector:
        matchLabels:
          access: "true"
```

```
$ kbtest apply -f test-policy.yaml --force
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x21de2bb]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/kubectl/resource.(*Helper).DeleteWithOptions(0xc4205dea00, 0x2c91e67, 0x7, 0xc420333ef0, 0xc, 0x0, 0xc4205dea00, 0x2c95017)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/helper.go:103 +0x3b
k8s.io/kubernetes/pkg/kubectl/resource.(*Helper).Delete(0xc4205dea00, 0x2c91e67, 0x7, 0xc420333ef0, 0xc, 0xc420186b80, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/helper.go:99 +0x5c
k8s.io/kubernetes/pkg/kubectl/cmd.runDelete(0x2c91e67, 0x7, 0xc420333ef0, 0xc, 0xc4201208c0, 0x0, 0x0, 0xc4202ee9c0, 0x1, 0xffffffffffffffff, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:523 +0x221
k8s.io/kubernetes/pkg/kubectl/cmd.(*patcher).delete(0xc420121570, 0x2c91e67, 0x7, 0xc420333ef0, 0xc, 0x1a6, 0xc420642f70)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:536 +0x9c
k8s.io/kubernetes/pkg/kubectl/cmd.(*patcher).deleteAndCreate(0xc420121570, 0xc4203f1900, 0x26e, 0x26e, 0x2c91e67, 0x7, 0xc420333ef0, 0xc, 0x2c91e67, 0x7, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:623 +0x63
k8s.io/kubernetes/pkg/kubectl/cmd.(*patcher).patch(0xc420121570, 0x3a44dc0, 0xc42000d730, 0xc4203f1900, 0x26e, 0x26e, 0xc4202c8e10, 0x10, 0x2c91e67, 0x7, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:617 +0x43a
k8s.io/kubernetes/pkg/kubectl/cmd.RunApply.func1(0xc4202d8980, 0x0, 0x0, 0x9, 0x21e7c01)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:320 +0x76a
k8s.io/kubernetes/pkg/kubectl/resource.ContinueOnErrorVisitor.Visit.func1(0xc4202d8980, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:348 +0x153
k8s.io/kubernetes/pkg/kubectl/resource.DecoratedVisitor.Visit.func1(0xc4202d8980, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:324 +0xdf
k8s.io/kubernetes/pkg/kubectl/resource.FlattenListVisitor.Visit.func1(0xc4202d8980, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:389 +0x572
k8s.io/kubernetes/pkg/kubectl/resource.EagerVisitorList.Visit.func1(0xc4202d8980, 0x0, 0x0, 0xe0, 0xc4202c8e10)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:207 +0x153
k8s.io/kubernetes/pkg/kubectl/resource.(*StreamVisitor).Visit(0xc4204ab540, 0xc4208eaec0, 0x3a4f1c0, 0xc4208eaf40)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:552 +0x350
k8s.io/kubernetes/pkg/kubectl/resource.(*FileVisitor).Visit(0xc4208ead00, 0xc4208eaec0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:501 +0x1a6
k8s.io/kubernetes/pkg/kubectl/resource.EagerVisitorList.Visit(0xc4201cf920, 0x1, 0x1, 0xc4208e86c0, 0x1, 0xc4208e86c0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:211 +0x100
k8s.io/kubernetes/pkg/kubectl/resource.(*EagerVisitorList).Visit(0xc4208eada0, 0xc4208e86c0, 0x4132000, 0x0)
	<autogenerated>:115 +0x69
k8s.io/kubernetes/pkg/kubectl/resource.FlattenListVisitor.Visit(0x3a4b240, 0xc4208eada0, 0xc4204ab500, 0xc4204ab5c0, 0xc4208eae01, 0xc4204ab5c0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:417 +0xa3
k8s.io/kubernetes/pkg/kubectl/resource.(*FlattenListVisitor).Visit(0xc4208eadc0, 0xc4204ab5c0, 0x18, 0x18)
	<autogenerated>:130 +0x69
k8s.io/kubernetes/pkg/kubectl/resource.DecoratedVisitor.Visit(0x3a4b2c0, 0xc4208eadc0, 0xc4208eae00, 0x3, 0x4, 0xc4208eae80, 0x1, 0xc4208eae80)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:325 +0xd8
k8s.io/kubernetes/pkg/kubectl/resource.(*DecoratedVisitor).Visit(0xc4208e8630, 0xc4208eae80, 0x100ca08, 0x2a737c0)
	<autogenerated>:151 +0x73
k8s.io/kubernetes/pkg/kubectl/resource.ContinueOnErrorVisitor.Visit(0x3a4b1c0, 0xc4208e8630, 0xc42009b6b0, 0x4132000, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:352 +0xf1
k8s.io/kubernetes/pkg/kubectl/resource.(*ContinueOnErrorVisitor).Visit(0xc4201cf940, 0xc42009b6b0, 0x100f2e8, 0xb0)
	<autogenerated>:144 +0x60
k8s.io/kubernetes/pkg/kubectl/resource.(*Result).Visit(0xc420132e70, 0xc42009b6b0, 0x0, 0xc4208e8690)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/resource/result.go:96 +0x62
k8s.io/kubernetes/pkg/kubectl/cmd.RunApply(0x3a708c0, 0xc4204cae70, 0xc420348fc0, 0x3a4a140, 0xc42000c018, 0x3a4a140, 0xc42000c020, 0xc420247b00, 0x2c8e585, 0x4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:339 +0x699
k8s.io/kubernetes/pkg/kubectl/cmd.NewCmdApply.func1(0xc420348fc0, 0xc4206ed1d0, 0x0, 0x5)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/apply.go:108 +0x188
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc420348fc0, 0xc4206ed180, 0x5, 0x5, 0xc420348fc0, 0xc4206ed180)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:603 +0x22b
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42040e240, 0x7000105, 0x0, 0xffffffffffffffff)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:689 +0x339
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42040e240, 0xc4204cae70, 0x3a4a100)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:648 +0x2b
k8s.io/kubernetes/cmd/kubectl/app.Run(0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubectl/app/kubectl.go:39 +0xd5
main.main()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:26 +0x22
```

--- a/pkg/apis/policy/validation/validation.go
+++ b/pkg/apis/policy/validation/validation.go
@@ -38,7 +38,7 @@ func ValidatePodDisruptionBudgetUpdate(pdb, oldPdb *policy.PodDisruptionBudget) 
 	restoreGeneration := pdb.Generation
 	pdb.Generation = oldPdb.Generation
 
-	if !reflect.DeepEqual(pdb, oldPdb) {
+	if !reflect.DeepEqual(pdb.Spec, oldPdb.Spec) {
 		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to poddisruptionbudget spec are forbidden."))
 	}
 	allErrs = append(allErrs, ValidatePodDisruptionBudgetStatus(pdb.Status, field.NewPath("status"))...)

--- a/pkg/apis/policy/validation/validation_test.go
+++ b/pkg/apis/policy/validation/validation_test.go
@@ -113,3 +113,111 @@ func TestValidatePodDisruptionBudgetStatus(t *testing.T) {
 		}
 	}
 }
+
+func TestValidatePodDisruptionBudgetUpdate(t *testing.T) {
+	c1 := intstr.FromString("10%")
+	c2 := intstr.FromInt(1)
+	c3 := intstr.FromInt(2)
+	oldPdb := &policy.PodDisruptionBudget{}
+	pdb := &policy.PodDisruptionBudget{}
+	testCases := []struct {
+		generations []int64
+		name        string
+		specs       []policy.PodDisruptionBudgetSpec
+		status      []policy.PodDisruptionBudgetStatus
+		ok          bool
+	}{
+		{
+			name:        "only update status",
+			generations: []int64{int64(2), int64(3)},
+			specs: []policy.PodDisruptionBudgetSpec{
+				{
+					MinAvailable:   &c1,
+					MaxUnavailable: &c2,
+				},
+				{
+					MinAvailable:   &c1,
+					MaxUnavailable: &c2,
+				},
+			},
+			status: []policy.PodDisruptionBudgetStatus{
+				{
+					PodDisruptionsAllowed: 10,
+					CurrentHealthy:        5,
+					ExpectedPods:          2,
+				},
+				{
+					PodDisruptionsAllowed: 8,
+					CurrentHealthy:        5,
+					DesiredHealthy:        3,
+				},
+			},
+			ok: true,
+		},
+		{
+			name:        "only update pdb spec",
+			generations: []int64{int64(2), int64(3)},
+			specs: []policy.PodDisruptionBudgetSpec{
+				{
+					MaxUnavailable: &c2,
+				},
+				{
+					MinAvailable:   &c1,
+					MaxUnavailable: &c3,
+				},
+			},
+			status: []policy.PodDisruptionBudgetStatus{
+				{
+					PodDisruptionsAllowed: 10,
+				},
+				{
+					PodDisruptionsAllowed: 10,
+				},
+			},
+			ok: false,
+		},
+		{
+			name:        "update spec and status",
+			generations: []int64{int64(2), int64(3)},
+			specs: []policy.PodDisruptionBudgetSpec{
+				{
+					MaxUnavailable: &c2,
+				},
+				{
+					MinAvailable:   &c1,
+					MaxUnavailable: &c3,
+				},
+			},
+			status: []policy.PodDisruptionBudgetStatus{
+				{
+					PodDisruptionsAllowed: 10,
+					CurrentHealthy:        5,
+					ExpectedPods:          2,
+				},
+				{
+					PodDisruptionsAllowed: 8,
+					CurrentHealthy:        5,
+					DesiredHealthy:        3,
+				},
+			},
+			ok: false,
+		},
+	}
+
+	for i, tc := range testCases {
+		oldPdb.Spec = tc.specs[0]
+		oldPdb.Generation = tc.generations[0]
+		oldPdb.Status = tc.status[0]
+
+		pdb.Spec = tc.specs[1]
+		pdb.Generation = tc.generations[1]
+		oldPdb.Status = tc.status[1]
+
+		errs := ValidatePodDisruptionBudgetUpdate(oldPdb, pdb)
+		if tc.ok && len(errs) > 0 {
+			t.Errorf("[%d:%s] unexpected errors: %v", i, tc.name, errs)
+		} else if !tc.ok && len(errs) == 0 {
+			t.Errorf("[%d:%s] expected errors: %v", i, tc.name, errs)
+		}
+	}
+}

